File: model/game_model.cpp
/* Jerry - A Chess Graphical User Interface
 * Copyright (C) 2014-2016 Dominik Klein
 * Copyright (C) 2015-2016 Karl Josef Klein
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include "game_model.h"
#include "chess/game.h"
#include "chess/board.h"
#include <QDebug>
#include <QPixmap>
#include <assert.h>
#include <QSettings>
#include <iostream>
#include "chess/pgn_printer.h"
#include "chess/pgn_reader.h"
#include <QtGui>
#include "internalengine.h"
#include "various/resource_finder.h"

GameModel::GameModel(QObject *parent) :
    QObject(parent)
{
    this->modelVersion = 3;
    this->wasSaved = false;
    this->lastSaveFilename = "";
    this->lastOpenDir = "";
    this->lastSaveDir = "";
    this->game = new chess::Game();   
    this->colorStyle = new ColorStyle(ResourceFinder::getPath());
    this->mode = MODE_ENTER_MOVES;
    this->engines = new QList<Engine*>();
    this->active_engine = new InternalEngine();
    this->engines->append(active_engine);
    this->lastAddedEnginePath = QString("");
    this->humanPlayerColor = chess::WHITE;
    // stockfish specific: 0 (min), 20 (max)
    this->engineStrength = 20;
    this->engineThinkTimeMs = 3000;
    this->flipBoard = false;

    this->currentBestPv = QString("");
    this->currentMateInMoves = -1;
    this->currentEval = -10000;

    this->prevBestPv = QString("");
    this->prevMateInMoves = -1;
    this->prevEval = -10000;

    this->analysisThreshold = 0.5;

    this->showEval = true;

    this->gameAnalysisStarted = false;

    this->gameAnalysisForPlayer = ANALYSE_BOTH_PLAYERS;

    this->company = QString("dkl");
    this->appId = QString("jerry_").append(JERRY_VERSION);

    QString temp = QString("mydatabase.dci");
    this->database = new chess::Database(temp);

    this->loadOpeningBook();
}

void GameModel::loadOpeningBook() {
    QString path = ResourceFinder::getPath().append("/books/");
    path = path.append(QString("varied.bin"));
    path = QDir::toNativeSeparators(QDir::cleanPath(path));
    this->book = new chess::Polyglot(path);
}

bool GameModel::canAndMayUseBook(chess::GameNode *node) {

    if(node->getDepth() > 20) {
        return false;
    } else {
        if(this->engineStrength <= 5 && node->getDepth() <= 8) {
            return true;
        }
        if(this->engineStrength <= 10 && node->getDepth() <= 12) {
            return true;
        }
        if(this->engineStrength > 10) {
            return true;
        }
        return false;
    }
}

QVector<chess::Move> GameModel::getBookMoves(chess::GameNode *node) {
    return this->book->findMoves(node->getBoard());
}

bool GameModel::isInBook(chess::GameNode *node) {
    return this->book->inBook(node->getBoard());
}

QString GameModel::getLastAddedEnginePath() {
    return this->lastAddedEnginePath;
}

void GameModel::setLastAddedEnginePath(QString &path) {
    this->lastAddedEnginePath = path;
}

void GameModel::setEngines(QList<Engine*> *engines) {
    this->engines = engines;
}

void GameModel::setActiveEngine(Engine *engine) {
    this->active_engine = engine;
}

chess::Game* GameModel::getGame() {
    return this->game;
}

void GameModel::setGame(chess::Game *g) {
    assert(g != 0);
    delete this->game;
    this->game = g;
    //this->game->setCurrent(this->game->getRootNode());
}

void GameModel::triggerStateChange() {
    emit(stateChange());
}

int GameModel::getMode() {
    return this->mode;
}

void GameModel::setMode(int mode) {
    this->mode = mode;
}

QList<Engine*> * GameModel::getEngines() {
    return this->engines;
}

Engine * GameModel::getActiveEngine() {
    return this->active_engine;
}

Engine* GameModel::getInternalEngine() {
    assert(this->engines->size() > 0);
    return this->engines->at(0);
}

int GameModel::getEngineThinkTime() {
    return 5;
}

int GameModel::getEngineStrength() {
    return this->engineStrength;
}

void GameModel::saveGameState() {

    QSettings settings(this->company, this->appId);

    settings.clear();

    settings.setValue("modelVersion", this->modelVersion);
    settings.setValue("lastSaveFilename", this->lastSaveFilename);
    settings.setValue("lastSaveDir", this->lastSaveDir);
    settings.setValue("lastOpenDir", this->lastOpenDir);

    settings.setValue("colorTheme", colorStyle->styleType);
    settings.setValue("pieceType", colorStyle->pieceType);
    settings.setValue("lastAddedEnginePath", this->lastAddedEnginePath);

    // stockfish specific settings
    settings.setValue("engineStrength", this->engineStrength);
    settings.setValue("engineThinkTimeMs", this->engineThinkTimeMs);
    settings.setValue("analysisThreshold", this->analysisThreshold);
    settings.setValue("showEval", this->showEval);

    settings.beginWriteArray("engines");
    for(int i=0;i<this->engines->size();i++) {
        settings.setArrayIndex(i);
        Engine *e = this->engines->at(i);
        settings.setValue("engineName", e->getName());
        settings.setValue("enginePath", e->getPath());
        if(e->isInternalEngine()) {
            settings.setValue("internalEngine", true);
        } else {
            settings.setValue("internalEngine", false);
        }
        if(e == this->active_engine) {
            settings.setValue("activeEngine", true);
        } else {
            settings.setValue("activeEngine", false);
        }
        settings.beginWriteArray("engineOptions");
        for(int j=0;j<e->getUciOptions()->size();j++) {
            settings.setArrayIndex(j);
            EngineOption *o = e->getUciOptions()->at(j);
            QString optNr = QString::number(j);
            QString option = optNr.append("option");
            QString val = optNr.append("value");
            QString type = optNr.append("type");
            settings.setValue(type, o->type);
            settings.setValue(option, o->toUciOptionString());
            if(o->type == EN_OPT_TYPE_CHECK) {
                settings.setValue(val, o->check_val);
            } else if(o->type == EN_OPT_TYPE_COMBO) {
                settings.setValue(val, o->combo_val);
            } else if(o->type == EN_OPT_TYPE_SPIN) {
                settings.setValue(val, o->spin_val);
            } else if(o->type == EN_OPT_TYPE_STRING) {
                settings.setValue(val, o->string_val);
            }
        }
        settings.endArray();
    }
    settings.endArray();

    // write current game as pgn to settings
    // first set dummy value
    settings.setValue("currentGame", "");
    chess::PgnPrinter *printer = new chess::PgnPrinter();
    try {
        QStringList* pgnStringList = printer->printGame(this->getGame());
        QString pgn = pgnStringList->join("\n");
        settings.setValue("currentGame", pgn);
        delete pgnStringList;
    } catch(std::exception e) {
        std::cerr << e.what() << std::endl;
    }
    delete printer;
    settings.sync();
}

void GameModel::restoreGameState() {

    QSettings settings(this->company, this->appId);

    if(settings.contains("currentGame")) {
        QString pgnString = settings.value("currentGame").toString();
        chess::PgnReader *reader = new chess::PgnReader();
        if(this->game != 0) {
            delete this->game;
        }
        try {
            this->game = reader->readGameFromString(&pgnString);
            this->game->findEco();
        } catch(std::exception e) {
            this->game = new chess::Game();
            std::cerr << e.what() << std::endl;
        }
        delete reader;
    }
    if(settings.contains("modalVersion")) {
        this->modelVersion = settings.value("modalVersion").toInt();
    }
    if(settings.contains("lastSaveFilename")) {
        this->lastSaveFilename = settings.value("lastSaveFilename").toString();
    }
    if(settings.contains("lastSaveDir")) {
        this->lastSaveDir = settings.value("lastSaveDir").toString();
    }
    if(settings.contains("lastOpenDir")) {
        this->lastOpenDir = settings.value("lastOpenDir").toString();
    }
    if(settings.contains("colorTheme")) {
        int styleType = settings.value("colorTheme").toInt();
        this->colorStyle->setStyle(styleType);
    }
    if(settings.contains("pieceType")) {
        this->colorStyle->pieceType = settings.value("pieceType").toInt();
    }
    if(settings.contains("lastAddedEnginePath")) {
        this->lastAddedEnginePath = settings.value("lastAddedEnginePath").toString();
    }
    if(settings.contains("engineStrength")) {
        this->engineStrength = settings.value("engineStrength").toInt();
    }
    if(settings.contains("engineThinkTimeMs")) {
        this->engineThinkTimeMs = settings.value("engineThinkTimeMs").toInt();
    }
    if(settings.contains("analysisThreshold")) {
        this->analysisThreshold = settings.value("analysisThreshold").toFloat();
    }
    /* omit show_eval: on starting up, eval should always
     * be displayed, no matter what user did before
    if(settings.contains("showEval")) {
        this->showEval = settings.value("showEval").toBool();
    }*/
    int size = settings.beginReadArray("engines");
    for(int i=0;i<size;i++) {
        settings.setArrayIndex(i);
        Engine* e = new Engine();
        if(settings.contains("engineName")) {
            QString eName = settings.value("engineName").toString();
            e->setName(eName);
        }
        if(settings.contains("enginePath")) {
            QString ePath = settings.value("enginePath").toString();
            e->setPath(ePath);
        }
        bool isInternal = false;
        if(settings.contains("internalEngine")) {
            isInternal = settings.value("internalEngine").toBool();
        }
        if(settings.contains("activeEngine")) {
            bool isActive = settings.value("activeEngine").toBool();
            if(isActive && !isInternal) {
                this->active_engine = e;
            }
        }
        int sizeOpts = settings.beginReadArray("engineOptions");
        for(int j=0;j<sizeOpts;j++) {
            settings.setArrayIndex(j);
            EngineOption *o = new EngineOption();
            QString optNr = QString::number(j);
            QString option = optNr.append("option");
            QString val = optNr.append("value");
            QString type = optNr.append("type");
            bool restored = false;
            if(settings.contains(option) && settings.contains(val) && settings.contains(type)) {
                int typeCode = settings.value(type).toInt();
                QString enopt = settings.value(option).toString();
                if(o->restoreFromString(enopt)) {
                    if(typeCode == EN_OPT_TYPE_CHECK) {
                        o->check_val = settings.value(val).toBool();
                        restored = true;
                    } else if(typeCode == EN_OPT_TYPE_COMBO) {
                        o->combo_val = settings.value(val).toString();
                        restored = true;
                    } else if(typeCode == EN_OPT_TYPE_SPIN) {
                        o->spin_val = settings.value(val).toInt();
                        restored = true;
                    } else if(typeCode == EN_OPT_TYPE_STRING) {
                        o->string_val = settings.value(val).toString();
                        restored = true;
                    }
                }
            }
            if(restored) {
                if(isInternal) {
                    Engine *internal = this->engines->at(0);
                    int idx = internal->existsEngineOption(o->name);
                    if(idx == -1) {
                        internal->getUciOptions()->append(o);
                    } else {
                        EngineOption *ati = internal->getUciOptions()->at(i);
                        delete ati;
                        internal->getUciOptions()->removeAt(i);
                        internal->getUciOptions()->append(o);
                    }
                } else {
                    e->getUciOptions()->append(o);
                }
            } else {
                delete o;
            }
        }
        settings.endArray();
        if(isInternal) {
            delete e;
        } else {
            this->engines->append(e);
        }
    }
    settings.endArray();
}

File: src/main/java/org/asdfjkl/jfxchess/gui/GameModel.java
/* JFXChess - A Chess Graphical User Interface
 * Copyright (C) 2020-2025 Dominik Klein
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.asdfjkl.jfxchess.gui;

import javafx.scene.control.Alert;
import javafx.stage.Stage;
import org.asdfjkl.jfxchess.lib.*;

import java.io.File;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.prefs.Preferences;
import java.lang.System;

import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class GameModel {

    public static final int MAX_PV = 64;
    public static final int MAX_N_ENGINES = 10;
    public static final int MODE_ENTER_MOVES = 0;
    public static final int MODE_ANALYSIS = 1;
    public static final int MODE_PLAY_WHITE = 2;
    public static final int MODE_PLAY_BLACK = 3;
    public static final int MODE_GAME_ANALYSIS = 4;
    public static final int BOTH_PLAYERS = 5;
    public static final int MODE_PLAYOUT_POSITION = 7;
    public static final int STYLE_LIGHT = 0;
    public static final int STYLE_DARK = 1;
    public int THEME = STYLE_DARK;
    Game game;
    private final ArrayList<StateChangeListener> stateChangeListeners = new ArrayList<>();
    private int currentMode;
    private boolean multiPvChanged = false;
    private boolean flipBoard = false;
    private boolean humanPlayerColor = CONSTANTS.WHITE;
    public boolean wasSaved = false;
    private int engineStrength = 2400;
    private int engineThinkTimeSecs = 3;

    ArrayList<Engine> engines = new ArrayList<>();
    Engine activeEngine = null;
    Engine selectedPlayEngine = null;
    Engine selectedAnalysisEngine = null;

    ArrayList<BotEngine> botEngines = new ArrayList<>();

    private int gameAnalysisForPlayer = BOTH_PLAYERS;
    private double gameAnalysisThreshold = 0.5; // pawns
    private int gameAnalysisThinkTimeSecs = 3;  // seconds

    private boolean gameAnalysisJustStarted = false;

    // to make sure that if the user play against the computer/bot
    // he is not able to/does not accidentally move the computer's pieces
    // on the computer's turn
    public boolean blockGUI = false;

    public String currentBestPv = "";
    public int currentBestEval = 0;
    public int currentMateInMoves = -1;
    public boolean currentIsMate = false;

    public String childBestPv = "";
    public int childBestEval = 0;
    public int childMateInMoves = -1;
    public boolean childIsMate = false;

    //public boolean doNotNotifyAboutResult = false;

    public PolyglotExt extBook;

    private Preferences prefs;

    private static final int modelVersion = 450;

    private final PgnDatabase pgnDatabase;
    public int currentPgnDatabaseIdx = -1;
    public File lastOpenedDirPath = null;
    public File lastSaveDirPath = null;

    public boolean openDatabaseOnNextDialog = false;

    public String extBookPath;

    public int maxCpus = 1;

    private SearchPattern searchPattern;
    BoardStyle boardStyle;

    Stage refToCurrentStage;

    public GameModel() {
        this.game = new Game();
        Board b = new Board(true);

        pgnDatabase = new PgnDatabase();
        searchPattern = new SearchPattern();
        searchPattern.setSearchForHeader(true);
        boardStyle = new BoardStyle();

        this.game.getRootNode().setBoard(b);
        this.currentMode = MODE_ENTER_MOVES;

        String stockfishPath = getStockfishPath();
	    // System.out.println("Computed Stockfish Path as: "+stockfishPath);

        Engine stockfish = new Engine();
        stockfish.setName(CONSTANTS.INTERNAL_ENGINE_NAME);
        if(stockfishPath != null) {
            stockfish.setPath(stockfishPath);
        }
        // System.out.println("stock fish path is now: "+stockfish.getPath());
        stockfish.setInternal(true);
        engines.add(stockfish);
        selectedAnalysisEngine = stockfish;
        activeEngine = stockfish;

        // manually add for internal stockfish up to 4 mpv
        EngineOption internalMPV = new EngineOption();
        internalMPV.name = "MultiPV";
        internalMPV.spinMin = 1;
        internalMPV.spinMax = 4;
        internalMPV.spinDefault = 1;
        internalMPV.spinValue = 1;
        internalMPV.type = EngineOption.EN_OPT_TYPE_SPIN;

        EngineOption internalElo = new EngineOption();
        internalElo.name = "UCI_Elo";
        internalElo.spinMin = 1320;
        internalElo.spinMax = 3190;
    	internalElo.spinDefault = 1320;
        internalElo.spinValue = 3190;
        internalElo.type = EngineOption.EN_OPT_TYPE_SPIN;

        EngineOption internalLimitStrength = new EngineOption();
        internalLimitStrength.name = "UCI_LimitStrength";
        internalLimitStrength.type = EngineOption.EN_OPT_TYPE_CHECK;
        internalLimitStrength.checkStatusDefault = false;
        internalLimitStrength.checkStatusValue = false;

        EngineOption internalThreads = new EngineOption();
        internalThreads.name = "Threads";
        internalThreads.type = EngineOption.EN_OPT_TYPE_SPIN;
        internalThreads.spinMin = 1;
        internalThreads.spinMax = 1024;
        internalThreads.spinDefault = 1;
        internalThreads.spinValue = 1;

        activeEngine.options.add(internalElo);
        activeEngine.options.add(internalMPV);
        activeEngine.options.add(internalLimitStrength);
        activeEngine.options.add(internalThreads);

        // add bots
        String botPath = getBotEnginePath();
        botEngines = BotEngines.createEngines(botPath);
        selectedPlayEngine = botEngines.get(0); // set benny as default; todo: remember last selected bot

        // System.out.println("stock fish path at the end of gamemodel: "+stockfish.getPath());
    }

    public void loadExtBook() {

        File file = null;
        extBook = new PolyglotExt();
        if(extBookPath.isEmpty()) {
            extBookPath = getExtBookPath();
        }
        if(extBookPath != null) {
            file = new File(extBookPath);
            extBook.loadBook(file);
        }

    }

    public void setStageRef(Stage stage) {
        refToCurrentStage = stage;
    }

    public Stage getStageRef() {
        return refToCurrentStage;
    }
    
    public Path getJarPath() {
    try {
    	Path jarPath = Paths.get(App.class.getProtectionDomain().getCodeSource().getLocation().toURI());
        Path jarDir = jarPath.getParent();
        // DialogSimpleAlert dlg = new DialogSimpleAlert(this.getStageRef(), Alert.AlertType.INFORMATION,
        //        "JAR PATH", jarDir.toString());
        //dlg.showAndWait();
        //System.out.println("GET JAR PATH: "+jarDir);
        return jarDir;
        } catch (URISyntaxException e) {
            System.err.println("[ERROR] Failed to resolve JAR location: " + e.getMessage());
            return null;
        } catch (Exception e) {
            System.err.println("[ERROR] Unexpected error locating engine binary: " + e.getMessage());
            return null;
        }
    }

    private String getStockfishPath() {

        // currently redundant, but let's keep it here if we
        // have different packaging requirements in the future
        Path jarDir = getJarPath();
        String os = System.getProperty("os.name").toLowerCase();
        if(os.contains("win")) {
            if (jarDir != null) {
                Path engineBinary = jarDir.resolve("engine").resolve("stockfish.exe");
                //System.out.println("WINDOWS Stockfish Internal@: "+engineBinary);
                //DialogSimpleAlert dlg = new DialogSimpleAlert(this.getStageRef(), Alert.AlertType.INFORMATION,
                //        "WIN STOCKFISH INTERNAL", engineBinary.toString());
                //dlg.showAndWait();
                return engineBinary.toString();
            }
        }
        if(os.contains("linux")) {
            Path engineBinary = null;
            if (jarDir != null) {
                engineBinary = jarDir.resolve("engine").resolve("stockfish_x64");
                //System.out.println("LINUX Stockfish Internal@: "+engineBinary);
                return engineBinary.toString();
            }
        }
        return null;
    }

    private String getBotEnginePath() {

        // currently redundant, but let's keep it here if we
        // have different packaging requirements in the future
        Path jarDir = getJarPath();
        String os = System.getProperty("os.name").toLowerCase();
        if(os.contains("win")) {
            if (jarDir != null) {
                Path engineBinary = jarDir.resolve("engine").resolve("stockfish5.exe");
                //System.out.println("WINDOWS Stockfish5 Internal@: "+engineBinary);
                //DialogSimpleAlert dlg = new DialogSimpleAlert(this.getStageRef(), Alert.AlertType.INFORMATION,
                //        "WIN STOCKFISH 5 INTERNAL", engineBinary.toString());
                return engineBinary.toString();
            }
        }
        if(os.contains("linux")) {
            Path engineBinary = null;
            if (jarDir != null) {
                engineBinary = jarDir.resolve("engine").resolve("stockfish5_x64");
                //System.out.println("LINUX Stockfish5 Internal@: "+engineBinary);
                return engineBinary.toString();
            }
        }
        return null;

    }

    public String getExtBookPath() {

        Path jarDir = getJarPath();
        String os = System.getProperty("os.name").toLowerCase();
        if(os.contains("win")) {
            Path extBookBinary = null;
            if (jarDir != null) {
                extBookBinary = jarDir.resolve("book").resolve("extbook.bin");
                //System.out.println("WINDOWS Book@: "+extBookBinary);
                //DialogSimpleAlert dlg = new DialogSimpleAlert(this.getStageRef(), Alert.AlertType.INFORMATION,
                //        "EXT BOOK BINARY", extBookBinary.toString());
                return extBookBinary.toString();
            }
        }
        if(os.contains("linux")) {
            Path extBookBinary = null;
            if (jarDir != null) {
                extBookBinary = jarDir.resolve("book").resolve("extbook.bin");
                //System.out.println("LINUX Book@: "+extBookBinary);
                return extBookBinary.toString();
            }
        }
        return null;
    }

    public Game getGame() {
        return game;
    }

    public void setGame(Game game) {
        this.game = game;
    }

    public int getGameAnalysisForPlayer() { return gameAnalysisForPlayer; }

    public void setGameAnalysisForPlayer(int player) { gameAnalysisForPlayer = player; }

    public double getGameAnalysisThreshold() { return gameAnalysisThreshold; }

    public void setGameAnalysisThreshold(double threshold) { gameAnalysisThreshold = threshold; }

    public void setGameAnalysisThinkTimeSecs(int thinktimeSecs) { gameAnalysisThinkTimeSecs = thinktimeSecs; }

    public int getGameAnalysisThinkTimeSecs() { return gameAnalysisThinkTimeSecs; }

    public void setMode(int mode) {
        this.currentMode = mode;
    }

    public void setComputerThinkTimeSecs(int secs) {
        engineThinkTimeSecs = secs;
    }

    public int getComputerThinkTimeSecs() {
        return engineThinkTimeSecs;
    }

    public int getEngineStrength() { return engineStrength; }

    public void setEngineStrength(int strength) { engineStrength = strength; }

    public int getMode() {
        return currentMode;
    }

    public int getMultiPv() {
        return activeEngine.getMultiPV();
    }

    public void setFlipBoard(boolean flipBoard) {
        this.flipBoard = flipBoard;
    }

    public boolean getFlipBoard() {
        return flipBoard;
    }

    public void setHumanPlayerColor(boolean humanPlayerColor) {
        this.humanPlayerColor = humanPlayerColor;
    }

    public boolean getHumanPlayerColor() {
        return humanPlayerColor;
    }

    public boolean getGameAnalysisJustStarted() { return gameAnalysisJustStarted; }

    public void setGameAnalysisJustStarted(boolean val) { gameAnalysisJustStarted = val; }

    public PgnDatabase getPgnDatabase() {
        return pgnDatabase;
    }

    public SearchPattern getSearchPattern() {
        return searchPattern;
    }

    public void setSearchPattern(SearchPattern searchPattern) {
        this.searchPattern = searchPattern;
    }

    public void setMultiPv(int multiPv) {
        if(multiPv >= 1 && multiPv <= activeEngine.getMaxMultiPV() && multiPv <= MAX_PV) {
            activeEngine.setMultiPV(multiPv);
            this.multiPvChanged = true;
        }
    }

    public boolean wasMultiPvChanged() {
        return this.multiPvChanged;
    }

    public void setMultiPvChange(boolean b) {
        this.multiPvChanged = b;
    }

    public void addListener(StateChangeListener toAdd) {
        stateChangeListeners.add(toAdd);
    }

    public void triggerStateChange() {
        for (StateChangeListener sl : stateChangeListeners)
            sl.stateChange();
    }

    public void saveModel() {

        prefs = Preferences.userRoot().node(this.getClass().getName());

        prefs.putInt("modelVersion",modelVersion);

        PgnPrinter printer = new PgnPrinter();
        String pgn = printer.printGame(getGame());
        prefs.put("currentGame", pgn);
    }

    public void savePaths() {
        prefs = Preferences.userRoot().node(this.getClass().getName());
        if(lastOpenedDirPath != null) {
            prefs.put("lastOpenDir", lastOpenedDirPath.toString());
        }
        if(lastSaveDirPath != null) {
            prefs.put("lastSaveDir", lastSaveDirPath.toString());
        }
    }

    public void saveBoardStyle() {
        prefs = Preferences.userRoot().node(this.getClass().getName());
        prefs.putInt("COLOR_STYLE", boardStyle.getColorStyle());
        prefs.putInt("PIECE_STYLE", boardStyle.getPieceStyle());
    }

    public void saveEngines() {

        prefs = Preferences.userRoot().node(this.getClass().getName());
        // Clean up preferences. Preferences for engines
        // which have been removed may still be there.
        for(int i=0;i<MAX_N_ENGINES;i++) {
            prefs.remove("ENGINE"+i);
        }
        for(int i=1;i<engines.size();i++) {
            Engine engine = engines.get(i);
            String engineString = engine.writeToString();
            prefs.put("ENGINE"+i, engineString);
        }
        prefs.putInt("ACTIVE_ENGINE_IDX", engines.indexOf(selectedAnalysisEngine));
    }

    public void saveExtBookPath() {
        prefs = Preferences.userRoot().node(this.getClass().getName());
        if(!extBookPath.isEmpty()) {
            prefs.put("EXT_BOOK_PATH_FILE", extBookPath);
        }
    }

    public void restoreExtBookPath() {

        prefs = Preferences.userRoot().node(this.getClass().getName());
        int mVersion = prefs.getInt("modelVersion", 0);

        String bookPath = getExtBookPath();
        if(mVersion == modelVersion) {
            bookPath = prefs.get("EXT_BOOK_PATH_FILE", bookPath);
        }
        extBookPath = bookPath;
    }

    public void saveGameAnalysisThresholds() {
        prefs = Preferences.userRoot().node(this.getClass().getName());
        prefs.putInt("GAME_ANALYSIS_SECS", getGameAnalysisThinkTimeSecs());
        prefs.putDouble("GAME_ANALYSIS_THRESHOLD", getGameAnalysisThreshold());
    }

    public void restoreGameAnalysisThresholds() {

        prefs = Preferences.userRoot().node(this.getClass().getName());
        int mVersion = prefs.getInt("modelVersion", 0);

        if(mVersion == modelVersion) {
            int gameAnalysisSecs = prefs.getInt("GAME_ANALYSIS_SECS", 3);
            double gameAnalysisThreshold = prefs.getDouble("GAME_ANALYSIS_THRESHOLD", 0.5);
            setGameAnalysisThinkTimeSecs(gameAnalysisSecs);
            setGameAnalysisThreshold(gameAnalysisThreshold);
        }
    }

    public void saveNewGameSettings() {
        prefs = Preferences.userRoot().node(this.getClass().getName());
        prefs.putInt("COMPUTER_THINK_TIME_SECS", getComputerThinkTimeSecs());
        prefs.putDouble("COMPUTER_STRENGTH", getEngineStrength());
    }

    public void restoreNewGameSettings() {

        prefs = Preferences.userRoot().node(this.getClass().getName());
        int mVersion = prefs.getInt("modelVersion", 0);

        if(mVersion == modelVersion) {
            int secs = prefs.getInt("COMPUTER_THINK_TIME_SECS", 3);
            int strength = prefs.getInt("COMPUTER_STRENGTH", 20);
            setComputerThinkTimeSecs(secs);
            setEngineStrength(strength);
        }
    }

    public void restoreBoardStyle() {

        BoardStyle style = new BoardStyle();
        prefs = Preferences.userRoot().node(this.getClass().getName());
        int mVersion = prefs.getInt("modelVersion", 0);

        if(mVersion == modelVersion) {
            int colorStyle = prefs.getInt("COLOR_STYLE", BoardStyle.STYLE_BLUE);
            int pieceStyle = prefs.getInt("PIECE_STYLE", BoardStyle.PIECE_STYLE_MERIDA);
            style.setPieceStyle(pieceStyle);
            style.setColorStyle(colorStyle);
        }
        boardStyle = style;
    }

    public void restoreEngines() {

        prefs = Preferences.userRoot().node(this.getClass().getName());
        int mVersion = prefs.getInt("modelVersion", 0);
        if (mVersion == modelVersion) {
            // don't restore engine with idx 0 (that's stockfish internal)
            for (int i = 1; i < MAX_N_ENGINES; i++) {
                String engineString = prefs.get("ENGINE" + i, "");
                if (!engineString.isEmpty()) {
                    Engine engine;
                    //if (i == 0) {
                    //    // engine 0 is Stockfish internal
                    //    // engine = engines.get(0);
                    //    // engine.restoreFromString(engineString);
                    //} else {
                        engine = new Engine();
                        engine.restoreFromString(engineString);
                        engines.add(engine);
                    //}
                }
            }
            int activeIdx = prefs.getInt("ACTIVE_ENGINE_IDX", 0);
            if(activeIdx < engines.size()) {
                System.out.println("RESTORE ACTIVE ENGINE IDX: "+activeIdx);
                activeEngine = engines.get(activeIdx);
                selectedAnalysisEngine = engines.get(activeIdx);
            } else {
                activeEngine = engines.get(0);
                selectedAnalysisEngine = engines.get(0);
            }
        }
    }

    public void saveScreenGeometry(ScreenGeometry g) {

        Preferences prefs = Preferences.userRoot().node(this.getClass().getName());

        prefs.putDouble("WINDOW_POSITION_X", g.xOffset);
        prefs.putDouble("WINDOW_POSITION_Y", g.yOffset);
        prefs.putDouble("WINDOW_WIDTH", g.width);
        prefs.putDouble("WINDOW_HEIGHT", g.height);
        prefs.putDouble("MOVE_DIVIDER_RATIO", g.moveDividerRatio);
        prefs.putDouble("MAIN_DIVIDER_RATIO", g.mainDividerRatio);
    }

    public ScreenGeometry restoreScreenGeometry() {

        prefs = Preferences.userRoot().node(this.getClass().getName());
        int mVersion = prefs.getInt("modelVersion", 0);

        ScreenGeometry g = new ScreenGeometry(-1.0, -1.0, -1.0,-1.0,
                ScreenGeometry.DEFAULT_MOVE_DIVIDER_RATIO,
                ScreenGeometry.DEFAULT_MAIN_DIVIDER_RATIO);

        if(mVersion == modelVersion) {

            g.xOffset = prefs.getDouble("WINDOW_POSITION_X", -1.0);
            g.yOffset = prefs.getDouble("WINDOW_POSITION_Y", -1.0);
            g.width = prefs.getDouble("WINDOW_WIDTH", -1.0);
            g.height = prefs.getDouble("WINDOW_HEIGHT", -1.0);
            g.moveDividerRatio = prefs.getDouble("MOVE_DIVIDER_RATIO",
                    ScreenGeometry.DEFAULT_MOVE_DIVIDER_RATIO);
            g.mainDividerRatio = prefs.getDouble("MAIN_DIVIDER_RATIO",
                    ScreenGeometry.DEFAULT_MAIN_DIVIDER_RATIO);

        }
        return g;

    }

    public void saveWindowMaxStatus(boolean isWindowMaximized) {
        Preferences prefs = Preferences.userRoot().node(this.getClass().getName());
        prefs.putBoolean("APP_WINDOW_MAX", isWindowMaximized);
    }

    public boolean restoreWindowMaxStatus() {
        Preferences prefs = Preferences.userRoot().node(this.getClass().getName());
        return prefs.getBoolean("APP_WINDOW_MAX", false);
    }

    public void saveToolbarVisibility(boolean isVisible) {
        Preferences prefs = Preferences.userRoot().node(this.getClass().getName());
        prefs.putBoolean("SHOW_TOOLBAR", isVisible);
    }

    public boolean restoreToolbarVisibility() {
        Preferences prefs = Preferences.userRoot().node(this.getClass().getName());
        return prefs.getBoolean("SHOW_TOOLBAR", true);
    }

    public void saveTheme() {
        Preferences prefs = Preferences.userRoot().node(this.getClass().getName());
        prefs.putInt("COLOR_THEME", this.THEME);
    }

    public void restoreTheme() {
        Preferences prefs = Preferences.userRoot().node(this.getClass().getName());
        this.THEME = prefs.getInt("COLOR_THEME", STYLE_DARK);
    }

    public void restorePaths() {
        prefs = Preferences.userRoot().node(this.getClass().getName());
        String lastOpenDir = prefs.get("lastOpenDir", "");
        String lastSaveDir = prefs.get("lastSaveDir", "");
        if (!lastOpenDir.isEmpty()) {
            lastOpenedDirPath = new File(lastOpenDir);
            if (!lastOpenedDirPath.exists()) {
                lastOpenedDirPath = null;
            }
        }
        if (!lastSaveDir.isEmpty()) {
            lastSaveDirPath = new File(lastSaveDir);
            if (!lastSaveDirPath.exists()) {
                lastSaveDirPath = null;
            }
        }
    }

    public void restoreModel() {

        prefs = Preferences.userRoot().node(this.getClass().getName());
        int mVersion = prefs.getInt("modelVersion", 0);

        if(mVersion == modelVersion) {
            PgnReader reader = new PgnReader();

            String pgn = prefs.get("currentGame", "");

            if(!pgn.isEmpty()) {
                Game g = reader.readGame(pgn);
                PgnPrinter p = new PgnPrinter();
                if (g.getRootNode().getBoard().isConsistent()) {
                    setGame(g);
                    g.setTreeWasChanged(true);
                }
            }
        }
    }

    public String getVersion() {

        int mainVersion = modelVersion / 100;
        int subVersion = (modelVersion % 100) / 10;
        int subSubVersion = ((modelVersion % 100) % 10);

        return mainVersion + "." + subVersion + "." + subSubVersion;

    }

}



